nntrainer/tensor/memory_pool.h

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file   memory_pool.h
+ * @date   10 August 2021
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is Memory Pool Class for
+ *
+ * @todo   Support an external allocator for different backends
+ */
+
+#ifndef __MEMORY_POOL_H__
+#define __MEMORY_POOL_H__
+
+#include <vector>
+
+#include <memory_planner.h>
+
+namespace nntrainer {
+
+/**
+ * @class   MemoryPool
+ * @brief   Memory Pool provides a common pool for all the tensor memory
+ */
+class MemoryPool {
+public:
+  /**
+   * @brief MemoryPool default constructor
+   *
+   */
+  MemoryPool() : mem_pool(nullptr), pool_size(0), min_pool_size(0) {}
+
+  /**
+   * @brief Request Memory from memory pool
+   *
+   * @param bytes The size of the memory requested in bytes
+   * @param start_time The start of the validity interval of this memory
+   * @param end_time The end of the validity interval of this memory
+   *
+   * @return The token to get the pointer for this memory after allocation
+   */
+  unsigned int requestMemory(size_t bytes, unsigned int start_time,
+                             unsigned int end_time);
+
+  /**
+   * @brief Planner the layout with memory planner
+   *
+   * @param planner The memory planner to be used for finalizing the layout
+   *
+   * @return The total memory requirement using this memory planner
+   * @return The efficiency of the memory layer with the given memory planner
+   *
+   * @details The efficiency of the planner is calculated as the ratio of the
+   * theoretical minimum memory requirement divided by the memory requirement
+   * given by the memory planner.
+   *
+   * @details planLayout can be called multiple times as this does not perform
+   * any allocation but rather just plans the layout and stores the layout.
+   * Subsequent call to this function will overwrite any existing layout.
+   */
+  double planLayout(const MemoryPlanner &planner);
```

Comment:
I guess this has to be pointer for polymorphism?

Reply:
MemoryPlanner is a stateless object. So, it can be const reference which can be created on fly.

Reply:
Yes, I mean to utilize virtual function shouldn't it be a pointer?

Reply:
i tested with BinaryPlanner and it works (check the next PR).

Reply:
Cool. Thanks :)